lib/ansible/modules/files/lineinfile.py

```diff
@@ -248,15 +252,37 @@ def present(module, dest, regexp, line, insertafter, insertbefore, create,
             try:
                 os.makedirs(b_destpath)
             except Exception as e:
-                module.fail_json(msg='Error creating %s Error code: %s Error description: %s' % (b_destpath, e[0], e[1]))
+                module.fail_json(msg='Error creating {0} Error code: {1} Error description: {2}'.format(b_destpath, e[0], e[1]))
 
-        b_lines = []
-    else:
+    with flock.lock_file(dest):
         with open(b_dest, 'rb') as f:
             b_lines = f.readlines()
```

Comment:
Once we have the flock.lock_file() code raising an IOError if the file does not exist, we can rewrite this section like this:
``` python
try:
    with flock.lock_file(dest):
        with open(b_dest, 'rb') as f:
            b_lines = f.readlines()
        msg, changed, b_modified_lines = _present_data_manipulator(b_lines, regexp, line, insertafter,
                                                                   insertbefore, backrefs, firstmatch)
        if changed and not module.check_mode:
            if backup and os.path.exists(b_dest):
                backupdest = module.backup_local(dest)
            write_changes(module, b_modified_lines, dest)
except IOError as e:
    if not module.check_mode and 'file doesn't exist when locking' in to_native(e):
        raise
    # Check mode won't create the file so we know it starts off empty
    b_lines = []
    msg, changed, b_modified_lines = _present_data_manipulator(b_lines, regexp, line, insertafter,
                                                                   insertbefore, backrefs, firstmatch)

diff['before'] = to_native(b('').join(b_lines))
[...]
```

One thing I don't know is whether check_file_attrs should be inside of the lock. If it does, I think it can still be added in up there but there would need to be further rearrangement or another test of whether we're in check_mode or not.

Reply:
I wrapped it in if/else, to prevent locking while running in check_mode, I think it's cleaner even though there's some duplicated code. I left check_file_attrs inside the lock to keep the results consistent.